src/libstore/gc.cc

```diff
@@ -786,8 +786,24 @@ void LocalStore::collectGarbage(const GCOptions & options, GCResults & results)
         for (auto & i : options.pathsToDelete) {
             assertStorePath(i);
             tryToDelete(state, i);
-            if (state.dead.find(i) == state.dead.end())
-                throw Error(format("cannot delete path '%1%' since it is still alive") % i);
+            if (state.dead.find(i) == state.dead.end()) {
+                /* We cannot delete the current path: it is referenced by
+                   at least an intermediate path linked to a GC root.
+                   Let's find this intermediate path. */
+                PathSet incomingLinks;
+                PathSet alreadyVisited;
+                Path gcRootLink;
+                queryReferrers(i, incomingLinks);
+                for (auto & j : incomingLinks)
+                    if(canReachRoot(state, alreadyVisited, i)) {
+                        gcRootLink = j;
+                        break;
+                    }
+                /* We are sure this path is linked to a GC root. We should
+                   have found something.*/
+                assertStorePath(gcRootLink);
+                throw Error(format("cannot delete path '%1%': it is still referenced by '%2%'") % i % gcRootLink);
```

Comment:
Maybe we should tell the user about `nix-store -q --roots` here.

Reply:
My general preference is rather than having a computer telling a user what to do, if at all possible it should just do that for the user. But perhaps that means just showing the output of `--roots` on a failure here.

Reply:
The thing is that the message is incomplete: it shows a single referrer, even though there might be many more. Also, the referrer might be irrelevant: you want to know what *roots* cause the path to be reachable.

Reply:
I would say that I can imagine (i.e. remember) all three situations: when a list of roots is the best explanation, when a list of referrers explains everything and the relevance of roots is unclear, and when the best explanation is a randomly selected route from a store path to a root.

Reply:
I agree.

Maybe I could write a pretty printer doing something like this:

```
├── $GC_ROOT
│   ├── $REF_1
│       ├── $REF_2
│           ├── $PATH
```

The remaining question being how do we handle multiple referrers? Do we just display one randomly? Do we successively display the different paths reading to a GC_ROOT?

In any case, it seems this issue is not as straightforward as I though...